### PR TITLE
Add a test case calling caffe2 layer_norm from caffe2 executor but through the c10 dispatcher

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#16283 Add a test case calling caffe2 layer_norm from caffe2 executor but through the c10 dispatcher**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13792591/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16570 Simplify layer_norm_op_test&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13883913/)

Differential Revision: [D13792591](https://our.internmc.facebook.com/intern/diff/D13792591/)

### CODE REVIEW

**Comment by ezyang** [comment]
I'm not sure where to put these comments, so I'm putting them at the top of the diff stack.

https://github.com/pytorch/pytorch/blob/9d43baed924aa65c83b8ad6e7e29ee80b45a3da0/aten/src/ATen/core/dispatch/Dispatcher.h#L54

Is there any situation we ever want to deregister a kernel? For the PyTorch-Caffe2 calling case, I don't think this is actually necessary, right?

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/blob/9d43baed924aa65c83b8ad6e7e29ee80b45a3da0/aten/src/ATen/core/dispatch/OpSchema.h#L14

Why does KernelState live in OpSchema? That doesn't make too much sense to me

**Comment by smessmer** [comment]
>  static void deregisterKernel(const typename Schema::dispatch::dispatch_key_type& dispatch_key) { 
> Is there any situation we ever want to deregister a kernel? For the PyTorch-Caffe2 calling case, I don't think this is actually necessary, right?

For caffe2 kernels probably not. It's reasonable enough to anticipate that we'll need it for any custom kernel libraries though, independent of the concrete design, so I'd keep it here. And actually, even for caffe2 kernels, it's cleaner to unload them at the end.

> Why does KernelState live in OpSchema? That doesn't make too much sense to me
good point, will try to move it somewhere else.

File: caffe2/python/operator_test/layer_norm_op_test.py [hunk 1]
```diff
@@ -117,6 +117,34 @@ def test_layer_norm_op(self, X, gc, dc):
             outputs_to_check=[0, 1, 2],
         )
 
+    @given(X=hu.tensors(n=1), **hu.gcs_cpu_only)
+    def test_layer_norm_op_c10(self, X, gc, dc):
+        X = X[0]
```
Comment (ezyang) [inline caffe2/python/operator_test/layer_norm_op_test.py]: Err... why would you do this. Just ask for `hu.tensor()` and get a single tensor in the first place.

File: caffe2/python/operator_test/layer_norm_op_test.py [hunk 2]
```diff
@@ -117,6 +117,34 @@ def test_layer_norm_op(self, X, gc, dc):
...
+        if len(X.shape) == 1:
+            X = np.expand_dims(X, axis=0)
```
Comment (ezyang) [inline caffe2/python/operator_test/layer_norm_op_test.py]: And if you wanted minimum dimensionality to be 2, pass `min_dim=2`

**Review by ezyang — APPROVED** [review]
with comments

**Comment by smessmer** [comment]
Created https://github.com/pytorch/pytorch/pull/16570 to fix your comments across all the test cases in here, not just the ones I added.

**Pull request closed without merging**